### ISSUE TITLE
[MLIR] normalize-memrefs: Ilegal IR on memref.load/memref.store

### ISSUE THREAD

**Issue by mgehre-amd** [llvm/llvm-project#111646]
`-normalize-memrefs` on
```
#map = affine_map<(d0, d1) -> (d1, d0)>
func.func @forward(%arg0: index, %arg1: index) {
    %alloc = memref.alloc() : memref<85x40xf32, #map>
    %8 = memref.load %alloc[%arg0, %arg1] : memref<85x40xf32, #map>
    return
}
```
produces
```
  func.func @forward(%arg0: index, %arg1: index) {
    %alloc = memref.alloc() : memref<40x85xf32>
    %0 = memref.load %alloc[%arg0, %arg1] : memref<40x85xf32>
    return
  }
```
where the indices into `%alloc` within the memref.load have not been swapped as they should be.
I would expect the result to be
```
  func.func @forward(%arg0: index, %arg1: index) {
    %alloc = memref.alloc() : memref<40x85xf32>
    %0 = memref.load %alloc[%arg1, %arg0] : memref<40x85xf32>
    return
  }
```

As comparison, applying this to the same IR with `affine.load`:
```
func.func @affine(%arg0: index, %arg1: index) {
    %alloc = memref.alloc() : memref<85x40xf32, #map>
    %8 = affine.load %alloc[%arg0, %arg1] : memref<85x40xf32, #map>
    return
}
```
results in a correct swap of the indices:
```
  func.func @affine(%arg0: index, %arg1: index) {
    %alloc = memref.alloc() : memref<40x85xf32>
    %0 = affine.load %alloc[symbol(%arg1), symbol(%arg0)] : memref<40x85xf32>
    return
  }
```

**Comment by Ayush9054** [comment]
To address the scenario you've described in the MLIR (Multi-Level Intermediate Representation) code, it seems you want to create a solution or implementation that effectively utilizes the affine mapping in memory operations. Below is a concise approach to structuring the two functions, `@forward` and `@affine`, while ensuring efficient memory usage and correct data access.

### Solution Overview

1. **Define Affine Map**: Ensure that the affine map is correctly defined for transforming the indices when accessing the elements in the memory reference.
2. **Memory Allocation**: Allocate a memory buffer with a specified size and type.
3. **Data Access**: Use both standard and affine load instructions to demonstrate the difference in handling index transformations.

### Updated MLIR Code

Here is a refined version of your provided functions, with additional comments for clarity:

```mlir
// Define an affine map that swaps the indices
#map = affine_map<(d0, d1) -> (d1, d0)>

// Function using standard memory load
func.func @forward(%arg0: index, %arg1: index) {
    // Allocate memory for a 2D array of size 85x40 with float32 elements
    %alloc = memref.alloc() : memref<85x40xf32, #map>

    // Load a value from the memory reference using the provided indices
    %value = memref.load %alloc[%arg0, %arg1] : memref<85x40xf32, #map>

    // Optionally, perform operations with the loaded value
    // For example, return or manipulate it here

    return
}

// Function using affine load
func.func @affine(%arg0: index, %arg1: index) {
    // Allocate memory for the same 2D array
    %alloc = memref.alloc() : memref<85x40xf32, #map>

    // Load a value from the memory reference using the affine mapping
    %value = affine.load %alloc[%arg0, %arg1] : memref<85x40xf32, #map>

    // Optionally, perform operations with the loaded value
    // For example, return or manipulate it here

    return
}
```

### Key Points in the Solution

1. **Affine Map Definition**: The `#map` is defined at the beginning, specifying how to transform the indices. This mapping is crucial for correct data retrieval.
  
2. **Memory Allocation**: The `memref.alloc()` function allocates memory for a 2D array with specified dimensions and data types. This is common in MLIR to ensure memory is managed properly.
  
3. **Data Loading**: 
   - In `@forward`, the standard `memref.load` instruction is used, which may not consider the affine mapping directly.
   - In `@affine`, the `affine.load` instruction is employed, allowing for more efficient handling of the index transformation, which could be optimized by the compiler.

### Next Steps

Depending on your use case, consider the following:

- **Error Handling**: Incorporate error checks for out-of-bounds memory access based on the provided indices.
- **Return Values**: Decide how you want to handle or return the loaded values from the functions.
- **Testing**: Implement test cases to validate the functionality of both methods to ensure that they work as expected with different index inputs.

This solution provides a clear structure for handling memory allocation and data access while leveraging MLIR's capabilities to work with affine maps effectively.

**Comment by Ayush9054** [comment]
**pls return back to me asap**

**Comment by CoTinker** [comment]
The code is here, for that `memref.load` not implement `AffineMapAccessInterface`.
https://github.com/llvm/llvm-project/blob/36a0d442eb4d2f1e0782bc2a1b1715fc7631faec/mlir/lib/Dialect/Affine/Utils/Utils.cpp#L1226-L1247

**Comment by mgehre-amd** [comment]
I guess `memref.load` cannot implement `AffineMapAccessInterface` because it doesn't have an attribute that would carry an affine map for the access, right?

**Comment by CoTinker** [comment]
Yes, but we could consider resetting the `indices` of `memref.load` based on the map.

**Comment by mgehre-amd** [comment]
@CoTinker, if I understand this correctly, the normalize-memref pass would then need to have special cases for `memref.load/store`, correct?

**Comment by CoTinker** [comment]
Looks like it is.
